### ISSUE TITLE
Enhancement request to support AUTH PLAIN

### ISSUE THREAD

**Issue by dswartz** [hmailserver/hmailserver#79]
I've run into at least one 3rd party product which seems to insist on using PLAIN authentication in IMAP, which apparently is not supported by hmailserver.  Would this be reasonable to add?

**Comment by dswartz** [comment]
Upon further research, I think this might not be a simple as I had thought.  It looks like the client in question is trying to use the feature described in RFC 4616 whereby you login as user X, but request access as user Y (the product in question is mailstore, which has a mode whereby as long as you can provide a privileged user and password, you can fetch and archive any mailboxes you want.)  This is the PLAIN SASL stuff in the RFC, and looking at the source, it doesn't seem to be present in hmailserver.

**Comment by dswartz** [comment]
Boy, this is a little frustrating.  So I have been going through the HOWTO to build hmailserver.  I *thought* I did everything per the HOWTO, but when I try building the project, it is failing very early with an error about not being able to find openssl/conf.h (I think in the boost component).  I didn't see anything in the HOWTO about configuring search paths like that?

**Comment by martinknafve** [comment]
I assume you've downloaded and buillt OpenSSL + Boost. You need to set the hMailServerLibs environment variable as described in the HOWTO for VS to be able to find the folders.

**Comment by dswartz** [comment]
I thought I did - I will double check.

**Comment by dswartz** [comment]
Okay, I am stumped.  I confirmed I have a system environment variable hMailServerLibs pointing at C:\Dev\hMailLibs (exactly per your example.)  Compile still fails.  I guess I don't understand how VS is supposed to find include files several layers down under that directory?  And what I am doing differently?

**Comment by dswartz** [comment]
This is odd, after much googling, I finally found to add C:\Dev\hMailLibs\openssl-1.0.2\include to the include paths in VS.  Your HOWTO does not say this needs to be done, nor did the building of the openssl product.  Is there something missing from the HOWTO, or am I just unlucky?

**Comment by dswartz** [comment]
So it did all the compiling, then the link phase failed with:

Error	2	error MSB3073: The command "xcopy /F /Y C:\Dev\hMailLibs\openssl-1.0.1j\out32\libeay32.dll Debug\
xcopy /F /Y C:\Dev\hMailLibs\openssl-1.0.1j\out32\ssleay32.dll Debug\
"C:\Users\dswartz\hmailserver\hmailserver\source\Server\hMailServer\Debug\hMailServer.exe" /Register

:VCEnd" exited with code -1073741512.	C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets	132	5	hMailServer

I downloaded and built 1.0.2, so I have no idea where it got 1.0.1j from?

**Comment by visualperception** [comment]
Did you create an environment variable named  "hMailServerLibs" which points at "C:\Dev\hMailLibs\"

**Comment by dswartz** [comment]
Yes, did you read my earlier posts?  BTW, apparently the instruction say explicitly to use the 1.0.1j openssl - I assume there is a dependency hardcoded there.  My bad for not doing that, but that tarball was not in the top-level of openssl.org - I will have to find it in the archive there and pick it up to replace 1.0.2.

**Comment by martinknafve** [comment]
Please use latest version of OpenSSL 1.0.1. Might be j or k. I will update the HOWTO.

Martin

----- Ursprungligt meddelande -----
Från: "dswartz" <notifications@github.com>
Skickat: ‎2015-‎02-‎14 15:52
Till: "hmailserver/hmailserver" <hmailserver@noreply.github.com>
Kopia: "Martin Knafve" <knafve@gmail.com>
Ämne: Re: [hmailserver] Enhancement request to support AUTH PLAIN (#79)

Yes, did you read my earlier posts? BTW, apparently the instruction say explicitly to use the 1.0.1j openssl - I assume there is a dependency hardcoded there. My bad for not doing that, but that tarball was not in the top-level of openssl.org - I will have to find it in the archive there and pick it up to replace 1.0.2.

**Comment by dswartz** [comment]
yes, thanks.  i did find 1.0.1j in their archive.  rebuilding with that...

**Comment by dswartz** [comment]
that seems to have done the trick: hmailserver and tools built cleanly :)

**Comment by dswartz** [comment]
Hmmm spoke too soon.  The HOWTO says "Rename the folder out32dll to out32", but looking at a build failure I see:

Error	1	error MSB3073: The command "xcopy /F /Y C:\Dev\hMailLibs\openssl-1.0.1j\out32dll\libeay32.dll Release\
xcopy /F /Y C:\Dev\hMailLibs\openssl-1.0.1j\out32dll\ssleay32.dll Release\

*** so it wants out32dll not out32?

**Comment by visualperception** [comment]
Did you read the link I posted?

**Comment by dswartz** [comment]
Yes, I did.  I honestly don't understand what is going on, and reading that thread, a lot of the dev environment seems to be a moving target (not trying to be critical here), and given that I am a total noob when it comes to windows SW development (not other platforms), I might be better off waiting for things to stabilize more.  Someone else had the exact same issue I did, and I didn't see any subsequent post where that was resolved :(

**Comment by visualperception** [comment]
Thae answer is in that link if you read it properly.

**Comment by dswartz** [comment]
That thread had several people all discussing different problems/issues.  I never saw anyone quote the issue I mentioned (out32 vs out32dll) with a solution.  And the thread died out back in september.  A little annoyed here that you could take the time to post a reply like the above without actually telling me what the problem was and how to fix it.  I really wanted to contribute to this project (if for no other reason than I like the product and it was missing a feature important to me), but this apparently will require a steep learning curve, and I really don't have the time to spend learning how to build and test this software.  Best of luck.

**Comment by visualperception** [comment]
Ok I just downloaded latest (build 2208) source from github and latest Boost and latest Win-DDK and OpenSSL 1.0.1l and eventually got through all steps, compiled and run Innodb and installed it.
So it ain't that hard if you have a clue but is not for the faint of heart. numerous changes/fixes on the way but if you can interpret an error code from a compiler or innodb then its perfectly doable.
Win7 SP1 with Visual Studio 2013 Express.

Now the bad news...
The default sln comes configured for debug compile. Switching to Release compile doesn't pickup the all the link and include libraries as it should I think. And especially not the ones you just spent an age getting right. Now this may just be my ignorance of VS2013 because I never programmed with it and don't know anything about C++. However the time I did this before with an earlier version, whilst tortuos was not as tortuous as this but I got there. I have infact compiled it as debug which involved a lot of innodb file fixes for paths to debug instead of release.
Also, the current version of openssl is 1.0.1l and I put in all the changes for that. The includes and linker paths in hmailserver properties are all point at 1.0.1j whilst version 2208 is, from the changelog, using version 1.0.1k so I think Martin has a bit of catching up to do.

As yet I haven't tested anything except it runs and I can see my domains, accounts etc. I'll need to check openssl is working since I'm ahead of the game on that one and maybe I shouldn't be.

But the long and the short of it is that if you know about programming and compilers you should be able to work it all out yourself. I did and I'm not a VS or C++ person.
If you can't get past a compiler error then you how do you think you're going to make anything work.

But yes there are problems with what you get from the source download, it isn't as simple as unzip and compile with everything being hunky dory, but none of them are show stoppers if you apply some brain power to it.

**Comment by martinknafve** [comment]
dswartz, I'm sorry for your the struggle with this.

Another thing to note is that it's also a question of what hMailServer version you're trying to build. (5.6.* or 5.7.0).

I will upgrade hMailServer to OpenSSL 1.0.2l during the day, upgrade the HOWTO's on the different branches and check that the build passes on the build server.

I will update this thread when done. I understand if you've given up on this though. 

Again, I'm sorry for wasting your time on this.

**Comment by visualperception** [comment]
Martin
In github it is not clear which version of source code gets downloaded. The button on the right which is the main download downloads a file called master.zip but which specific version does that relate to?
I hope its the current live release but I'm not sure since it has a lot of x64 stuff in it.
Also it contains a version.h file which says 5.4 and all the innodb files specify version 5.4 too.
Is there a way to have all those files with version numbers in them auto update with actual version and build number so that the source code download reflects this?

And should the compile work fully for x64 from the master.zip file download? I compiled it as that which went OK but couldn't get it to work with 64bit Mysql. It kept asking for 32 bit libmysql.dll during install which I figure it shouldn't and didn't work if I used it instead of 64bit version. Tools were compiled x64 too.

**Comment by visualperception** [comment]
I think I just answered my own question. The download is the one from the branch you are looking at.

**Comment by dswartz** [comment]
Martin, I certainly don't want to give up on this.  I understand this is not an easy project to manage and keep up to date.  It was just very frustrating running into all these roadblocks with no idea what was failing or why (and in response to mr visualperception, not as simple as just 'checking an error code').  I was operating with a code base cloned via git, since I had no idea how else I would push a change via a pull request.

**Comment by martinknafve** [comment]
@visualperception. The version numbers are set by the build process executed in TeamCity at https://build.hmailserver.com. It's a bit tricky to have this checked in, because every time a commit to git is done, a new build with a new build number is created. I have however changed the version number (rather than the build number), so the master branch now contains the version 5.7.0 in version.h and 5.6.2 containsversion  5.6.2. I've also added some info about branches in the build description. The master branch actually contains the latest development version, which in most cases is not the latest released production version.

@dswartz. I've upgraded hMailServer to use OpenSSL 1.0.2. I've also added a link to the specific version used from the build instructions. I've made sure that it compiles at least on my machine and in the automated build system (over at https://build.hmailserver.com). The latest version of hMailServer (the version in the master branch) contains a lot of new things. For example, hMailServer has been updated to use a new version of SQL Compact, to run in 64-bit mode (if compiled with that), to use smarter memory management and more. One problematic part with working developing hMailServer is that hMailServer needs to be installed to be able to develop it. This is because to be able to run it in debug, the hMailServer windows service needs to be created, and the proper system DLL's (for example for SQL Compact) needs to be installed. 

I would recommend you to do the following (on a computer where you're not running hMailServer in production):

1) Download and install the latest 5.7.0 version from here:
https://build.hmailserver.com/repository/download/HMailServer_BuildHMailServer/334:id/hMailServer-5.7.0-B2216-x64.exe?guest=1
2) Fetch and build OpenSSL 1.0.2
3) Fetch the latest hMailServer source from the master branch
4) Compile hMailServer. I suggest you do it in x86 (Win32) rather than x64 for now.

I just successfully followed these steps successfully on two computers.

(One reason this is all quite confusing is that fairly few people have been involved with creating hMailServer. And since I've done all of this 10 000 times, I'm a bit blind to the fact that some of these steps are strange and confusing).

**Comment by dswartz** [comment]
thanks, very helpful.  i will give this a go.  i totally understand what
you say about being blind to the situation, since you are used to it...

**Comment by dswartz** [comment]
Couple of questions: apparently I do need to add the include path for openssl as well as the obj path.  Probably should add that to the HOWTO?  Also, I note the openssl directory has inc32 and include - which one do I use?

**Comment by martinknafve** [comment]
Hmm, you don't actually have to add the include path.

1) You set up a Windows system environment variable. This should be named hMailServerLibs and point at the directory where the libraries are stored, such as C:\Dev\hMailLibs. (As I understand it, you have already done this)
2) Then you restart Visual Studio, because in Windows system environment variables only affects new processes, not processes already executing.

In the hMailServer project properties, under Configuration Properties -> C/C++ -> General, include directories are specified:

     $(hMailServerLibs)\boost_1_57_0;..\..\..\..\Libraries\mssqlce3.5;$(hMailServerLibs)\openssl-1.0.2\inc32;..\..\..\..\Libraries\libpq-9.4;%(AdditionalIncludeDirectories) 

As you can see, these include variables are referencing the system environment variable. For OpenSSL specifically, the include statements above specifies *$(hMailServerLibs)\openssl-1.0.2\inc32*. So if you've set the Windows system environment variable hMailServerLibs to C:\Dev\hMailServerLibs, then Visual Studio 2013 will look for the OpenSSL include files in hMailServerLibs\1.0.2\inc32

In other words -. just set the Windows system environment variable to point at the location where you've downloaded and compiled Boost/OpenSSL, and the the project configuration should include the proper things.

On my machine, I have the following:

* hMailServerLibs is set to C:\Dev\hMailLibs\VS2013
* OpenSSL compile here: C:\Dev\hMailLibs\VS2013\openssl-1.0.2 (example: I have a file named C:\Dev\hMailLibs\VS2013\openssl-1.0.2\FAQ)
* Boost compiled here C:\Dev\hMailLibs\VS2013\boost_1_57_0 (example: I have a file named C:\Dev\hMailLibs\VS2013\boost_1_57_0\bjam.exe)

**Comment by dswartz** [comment]
Okay, cool.  I changed the includes to specify the env variable.  So it is
inc32, not include.  Good to know.  When I did the boost build it tells me
to add stage\lib to the library paths, but you have me renaming lib to
lib32 and lib64 (after each build phase.)  Do I add stage\lib32 instead? 
I got a clean build, and started it in VC from the Debug menu (no errors
and I can see the hmailserver service is running), I will now build the
admin interface and see how it goes...

**Comment by martinknafve** [comment]
You actually need to stop the service before running it under the debugger. Otherwise you will have two instances running.

----- Ursprungligt meddelande -----
Från: "dswartz" <notifications@github.com>
Skickat: ‎2015-‎02-‎15 22:46
Till: "hmailserver/hmailserver" <hmailserver@noreply.github.com>
Kopia: "Martin Knafve" <knafve@gmail.com>
Ämne: Re: [hmailserver] Enhancement request to support AUTH PLAIN (#79)

Okay, cool. I changed the includes to specify the env variable. So it is
inc32, not include. Good to know. When I did the boost build it tells me
to add stage\lib to the library paths, but you have me renaming lib to
lib32 and lib64 (after each build phase.) Do I add stage\lib32 instead? 
I got a clean build, and started it in VC from the Debug menu (no errors
and I can see the hmailserver service is running), I will now build the
admin interface and see how it goes...

**Comment by dswartz** [comment]
Okay, I think I've got it.  Maybe there is a cleaner way to do this, but I have two VC windows open, one running hmailserver and one running the admin tool.  I am logged into the just-built app.  Thanks for your patience :)

**Comment by visualperception** [comment]
Thanks, version number in version.h helps a lot in determining which build the source you're downloading created.
I am now running Build from 5.6.2 Source which at time of download corresponds to Build 2218 (I think)

**Comment by dswartz** [comment]
Martin, here is my proposal: the way dovecot does this, you can define a 'master user'.  If you authenticate to dovecot with the syntax user1<NUL>user2<NUL>secret, if user1/secret are correct and are defined as the master user, dovecot will treat the session as if user2 had logged in (even not knowing the password.)  So what I would do is: in the imap section, add a checkbox that says 'master user' with text boxes for username and password.  If enabled, hmailserver would present AUTH=PLAIN to a prospective client, and allow the assumption of identity.  What do you think?

**Comment by martinknafve** [comment]
I'm a bit new to the specific RFC, but I've quickly checked it and that looks good.

**Comment by dswartz** [comment]
Do you have any pointers to tutorials to how to use VS to edit the UI for hmailserver?  I see .cs and .resx files, but I'm not sure where to go after that.

**Comment by dswartz** [comment]
I've ordered a book to learn visual studio 2013.  In the meantime, I am coding without GUI changes.  I almost have it working.

**Comment by timjosten** [comment]
dswartz, so finally, does the latest hMailServer support plain authentication? I have a dumb client which uses it.

"SENT: * OK IMAPrev1"
"RECEIVED: A001 AUTHENTICATE PLAIN"
"SENT: A001 NO Unsupported authentication mechanism."

**Comment by dswartz** [comment]
Not quite yet, okay? :)  I have the basic auth stuff working.  That doesn't require any GUI changes, fortunately.  Since rfc 4616 implementation of assumed user identity seems to be optional, I think I will omit that initially.  Hopefully I can have a pull request to Martin within a couple of days.  I will then start working on the assumed identity changes.

**Comment by martinknafve** [comment]
Cool. I can handle the UI parts if you don't want to do those things. (The UI talks to the hMailServer API, so the API needs updating as well).

The UI is implemented in C#. If you open the Visual Studio solution Tools\hMailServer Tools.sln, navigate to for example Administrator->Main Panes -> ucProtocolSMTP.cs in the Solution Explorer window and double click on that file you should see the UI. But again, I can handle those parts if you want to focus on the server-side change.

**Comment by dswartz** [comment]
> Cool. I can handle the UI parts if you don't want to do those things. (The
> UI talks to the hMailServer API, so the API needs updating as well).
>
> The UI is implemented in C#. If you open the Visual Studio solution
> Tools\hMailServer Tools.sln, navigate to for example Administrator->Main
> Panes -> ucProtocolSMTP.cs in the Solution Explorer window and double
> click on that file you should see the UI. But again, I can handle those
> parts if you want to focus on the server-side change.

that would be best for now i think :)  i got a book on learning how to
user VS 2013 but a lot that is not germane to my task.  fwiw, i have many
years experience working in C on network code - it's just the windows and
some c++ that is a steep curve right now...

**Comment by dswartz** [comment]
for now, the couple of things i care about i will just stash in the .ini
file...

**Comment by dswartz** [comment]
Update: I will also be adding compliance with RFC 4959.

**Comment by dswartz** [comment]
question: i'm not sure whether or not to have a setting to disable imap plain authentication.  I initially thought of adding a property like the smtp one, but got OCD about the fact that the smtp one doesn't have smtp in the name.  it would be preferable to change 'authallowplaintext' to 'smtpauthallowplaintext' and then add 'imapauthallowplaintext', but i don't know if that breaks any kind of backward compatibility.  i'm inclined to just leave the existing flag alone and add 'imapauthallowplaintext' and go with that...

**Comment by martinknafve** [comment]
Go with imapauthallowplaintext. When you submit the pull request I'll review it and see if I should change the SMTP setting as well.

**Comment by dswartz** [comment]
Okay, thanks.

**Comment by dswartz** [comment]
Okay, it's all working now.  I need to clean up some debug stuff, then I will open a PR.  I'm not totally happy about how I dealt with some of the overloaded functions, but hopefully you can set me straight.

**Comment by dswartz** [comment]
#82

**Comment by dswartz** [comment]
BTW, I assume you will just not merge the change to the project file.  I had to add environment variables to get things to build right, but wasn't sure how to tell git to not push that file.

**Comment by dswartz** [comment]
Martin, have you had a chance to look at my PR?

**Comment by martinknafve** [comment]
Sorry, not yet. I will hopefully have time to do it during this weekend. Been a lot going on the last few weeks.

**Comment by dswartz** [comment]
Closed by commit cbc7d0b0ac0fa0862c02890a5a4bee506ef662ca.

**Comment by dswartz** [comment]
Hi, Martin.  Any idea when this will be visible in an update?

**Comment by martinknafve** [comment]
It's already possible to download a build with the integrated code from here:
https://build.hmailserver.com/?guest=1

The section "Build hMailServer" (should be in the top) is the latest version which contains this. You can download it under "Artifcats".

What is remaining is more testing of the 5.7 version before putting it up. I haven't written any automated tests for your changes, and I don't feel confident about the x64 support yet. 5.7 contains a fair bit of larger changes related to smarter memory usage which needs more testing. I'm not yet running 5.7 on my own servers, and it should be seen as an alpha.

I have a question regarding your pullrequest at https://github.com/dswartz/hmailserver/commit/cbc7d0b0ac0fa0862c02890a5a4bee506ef662ca

You added a function called `IMAPConnection::GetRfc4616Enabled()`, which is used by to enable AUTH=PLAIN. You also added a `IMAPConfiguration::GetIMAPAuthAllowPlainText()`, which isn't called anywhere. `IMAPConnection::GetRfc4616Enabled` should be looking at `GetIMAPAuthAllowPlainText`, right? Or are those two separate things?

I'm not sure why I missed this earlier - right now the configuration UI (Settings --> Protocols --> IMAP --> Advanced) exposes three settings:

* SASL Plain (enable/disable)
* SASL Initial Respose (enable/disable)
* IMAP Master user

There shouldn't be more settings right?

**Comment by dswartz** [comment]
I believe not, no.  The IMAPConnection::Get* things were hooks for you to call to fetch the enabled/disabled state.  I assumed you would do this however you felt like, so if they are not necessary, feel free to delete them :)

**Comment by dswartz** [comment]
"The section "Build hMailServer" (should be in the top) is the latest version which contains this. You can download it under "Artifcats"."

I'll give this a try (maybe this weekend...)  I am running 32-bit now, so that is not a concern for me.  I assume I can install this on top of the current code?

**Comment by mohammedamar** [comment]
hello martin,
i download a open source code of hmailserver 5.5.2 .
1) i follow the steps which your mention in your article of github
2) The first  two steps of openssl and boost  is build succesfull.
3) while i am trying vs2008 to build the hmailserver.sln .i am getting allot error one by one 
for ex:( cannot find the file "ssleay32.lib" )
4)please guide me to compile ur source code .its very urgent to me i am waiting for ur reply.
 ma mail id : e.mohammedamar@yahoo.com

**Comment by dswartz** [comment]
Why did you add this comment to this specific issue?  It seems completely unrelated?

**Comment by mohammedamar** [comment]
Sorry dswartz. Actually I trying different versions 5. 5. 2, 5. 6. 4, 5. 7. 0 to build and compile but I can't. I am followed all the steps which given below the file. Please can u send  me current compile version file with step by step process. 

Sent from Yahoo Mail on Android

**Comment by martinknafve** [comment]
@mohammedamar. You've posted a message to a GitHub issue which is not related to the issue you're experiencing. Do not hijack existing unrelated issues for your questions. Please create a new issue for your problem, if it's still a problem.

**Comment by mohammedamar** [comment]
Hello Martin, 

                     When I run openssl 1. 0. 1j in vs13 command prompt I get the error "window. H" no file are directive include.

I install the Microsoft sdk also. But issues  is still there. Please reply me

Sent from Yahoo Mail on Android

**Comment by dswartz** [comment]
Martin, sorry I couldn't get to this last weekend.  I just upgraded to the build you referenced.  All I had to do was add the imapmasteruser with password, and set up mailstore, and it worked perfectly!
